handler/src/main/java/io/netty/handler/ssl/PemX509Certificate.java

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import java.math.BigInteger;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.Principal;
+import java.security.PublicKey;
+import java.security.SignatureException;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateExpiredException;
+import java.security.cert.CertificateNotYetValidException;
+import java.security.cert.X509Certificate;
+import java.util.Date;
+import java.util.Set;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.base64.Base64;
+import io.netty.util.CharsetUtil;
+import io.netty.util.IllegalReferenceCountException;
+
+/**
+ * This is a special purpose implementation of a {@link X509Certificate} that can
+ * only be used in conjunction with {@link OpenSslContext} and it cannot be used
+ * to do crypto stuff in Java land.
+ *
+ * All methods other than what's implemented in {@link PemEncoded}'s throw
+ * {@link UnsupportedOperationException}s.
+ *
+ * @see PemEncoded
+ * @see OpenSslContext
+ */
+public final class PemX509Certificate extends X509Certificate implements PemEncoded {
+
+    private static final byte[] BEGIN_CERT = "-----BEGIN CERTIFICATE-----\n".getBytes(CharsetUtil.US_ASCII);
+    private static final byte[] END_CERT = "\n-----END CERTIFICATE-----\n".getBytes(CharsetUtil.US_ASCII);
+
+    /**
+     * Creates a {@link PemX509Certificate} from a {@link X509Certificate}.
+     */
+    static PemX509Certificate toPEM(X509Certificate cert) throws CertificateEncodingException {
+        if (cert instanceof PemX509Certificate) {
+            return (PemX509Certificate) cert;
+        }
+
+        ByteBuf encoded = Unpooled.wrappedBuffer(cert.getEncoded());
+        try {
+            ByteBuf pem = Unpooled.buffer();
+            pem.writeBytes(BEGIN_CERT);
+            ByteBuf base64 = Base64.encode(encoded, true);
```

Comment:
not introduced by this PR but if we adjust the ordering of events such that the encoding is done first ... we should know the exact size to allocate `pem` (`BEGIN_CERT.length + base64.readableBytes() + END_CERT.length`)